fix(sessions): Limit number of groups in sessions_v2 API [ISSUE-1372]

jjbayer opened the pull request:
Previously, the `/sessions/` api would return an unlimited number of data points, depending on the number of different groups existing in the dataset (e.g. for 3 releases and 2 environments, group by release and environment would return 6 groups).

Because time series are also grouped by timestamp, the number of data points would easily exceed the default snuba limit of 1000 rows. In this case, the api would simply return a zero or null value for any data point not present in the snuba result (see https://getsentry.atlassian.net/browse/ISSUE-1372).

This PR

1. Increases the limit on the snuba query to 5000.
2. Limits the number of groups returned for each time series such that only the top-N groups are returned.
    1. N is determined by the number of intervals selected through `statsPeriod` and `interval`
    1. Groups are ranked in descending order by the value of the first `field` requested, e.g. `-sum(session)`.

This implements Option D2 from [this](https://www.notion.so/sentry/Paginating-the-Sessions-V2-API-a635da3c91ba4a52a75ef705c2fcef97) document.

The same logic for the metrics-based release health backend is implemented in https://github.com/getsentry/sentry/pull/31465 and will be merged into this PR.

jjbayer committed:
wip: Initial implementation and integration test

jjbayer committed:
fix: Primary column

jjbayer committed:
ref(metrics): Ensure same number across sessions_v2 queries

jjbayer committed:
Merge branch 'ref/metrics-sessions-v2-groups' into fix/sessions-v2-limit

jjbayer committed:
Merge branch 'master' into fix/sessions-v2-limit

jjbayer commented on src/sentry/snuba/sessions_v2.py:
```diff
@@ -421,6 +425,14 @@ def _run_sessions_query(query):
+    # We only return the top-N groups, based on the first field that is being
+    # queried, assuming that those are the most relevant to the user.
+    # In a future iteration we might expose an `orderBy` query parameter.
+    orderby = [f"-{query.primary_column}"]
```
It seems a bit arbitrary to let the first specified field decide what groups will be returned, but I don't have a better idea.

ahmedetefy replied:
Hmm for some reason I was under the impression we were going to use the `totals` field for that?

jjbayer commented on src/sentry/snuba/sessions_v2.py:
```diff
@@ -421,6 +425,14 @@ def _run_sessions_query(query):
+    # In a future iteration we might expose an `orderBy` query parameter.
+    orderby = [f"-{query.primary_column}"]
+
+    max_groups = SNUBA_LIMIT // len(get_timestamps(query))
```
We talked about a fixed limit of 50, but this would not allow us to query more than 200 time intervals (50 * 200 = 10000), so I'd rather make it dynamic.

ahmedetefy replied:
I agree. I think this approach is better
I thought however you are developing this (initially atleast) for the specific use case in the UI where you request 50 releases and 90 data pts -> 50*90 = 4500, and we bump snuba limit to 10,000

jjbayer commented:
@ahmedetefy @untitaker @Swatinem I did not implement the metrics-based backend yet, but opened this up for review to discuss two items that I missed during our meeting (see comments attached to code):

* Clients can request more than one field, so the first field would have to define what the "Top-N" groups are, by convention.
* A limit of 50 groups is too high when the number of time intervals surpasses 200, so I am now in favor of applying a dynamic limit depending on the number of intervals.

jjbayer committed:
Merge branch 'master' into fix/sessions-v2-limit

jjbayer committed:
Merge remote-tracking branch 'origin/master' into fix/sessions-v2-limit

jjbayer committed:
fix: Apply redundant condition in series query

jjbayer committed:
fix: Prevent empty tuples

jjbayer committed:
Merge branch 'master' into fix/sessions-v2-limit

jjbayer committed:
fix: Revert spurious change

jjbayer commented on src/sentry/snuba/sessions_v2.py:
```diff
@@ -315,9 +317,10 @@ def __init__(self, query, params, allowed_resolution: AllowedResolution):
+#: of time intervals should never exceed that limit.
 #: https://github.com/getsentry/snuba/blob/69862db3ad224b48810ac1bb3001e4c446bf0aff/snuba/query/snql/parser.py#L908-L909
-SNUBA_LIMIT = 1000
+SNUBA_LIMIT = 10000
```
@evanh I agree with your comments that sending 10000 data points to a UI rarely makes sense. However, this endpoint is also public API used by external clients, which presumably expect a complete set of data. So I think returning as many groups as possible is the best we can do for that use case.

evanh replied:
I don't understand that reasoning. Returning less data points doesn't make the data incomplete, just less granular (e.g. daily instead of hourly).

If this endpoint is going to be a public API then we should be more restrictive by default, instead of less restrictive.

Finally while this will be a public API, it's purpose is to power our UI and we should design it with that in mind. Returning that many data points doesn't improve the UI and might even degrade performance since it has to discard/render a bunch of useless points.

ahmedetefy reviewed (commented):
This PR seems fine to me but I let @evanh have the final say!

jjbayer committed:
feat: Set snuba limit to 5000 and document

jjbayer committed:
Merge branch 'master' into fix/sessions-v2-limit

jjbayer committed:
fix(sessions): Limit number of groups in metrics_sessions_v2 (#31465)

Apply the logic described in #31392 to the metrics-based implementation
of the sessions api.
